Marketplace: don't query the product list on every navigation

gcsecsey opened the pull request:
#### Changes proposed in this Pull Request

* add a check to the `<QueryProductsList />` component to avoid re-fetching the data on each page load
* add a `type` field to the `productsList` state variable to track which type of list has been fetched

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Number of calls are reduced ([video](https://d.pr/v/Lm4XmR)):
* Open devtools, go to the Network tab, and filter for `products`
* Visit the `/plugins` page
* Visit any plugin page
* Navigate back to the plugin browser page, and perhaps to other plugin pages
* Verify that the `https://public-api.wordpress.com/rest/v1.1/products` request is only sent out once

Caching works correctly when different types are fetched:
* Go to a self-hosted jetpack site
* Delete the `productsList` key in indexedDB
* Visit `/plans/[DOMAIN]`
* Visit `/plugins/[DOMAIN]`
* Verify that there are no regressions

![image](https://user-images.githubusercontent.com/11555574/150160098-40fce8a5-cdc4-4ee2-8390-8d03411c416c.png)

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes #59899

gcsecsey committed:
Don't query the product list on every navigation

gcsecsey committed:
add null check

cpapazoglou commented on client/components/data/query-products-list/index.jsx:
```diff
@@ -1,10 +1,14 @@
 
 const request = ( props ) => ( dispatch, getState ) => {
-	if ( isProductsListFetching( getState() ) ) {
+	const productsList = getProductsList( getState() );
+	if (
+		isProductsListFetching( getState() ) ||
+		( productsList && Object.keys( productsList ).length > 0 )
+	) {
```
Instead of introducing such a big change for all components that rely on this (and might need fresh data), would it be safer if we added a prop to the component like `refetch: false/true` which by default would be `true` so that we keep backwards compatibility?

gcsecsey replied:
That's a great point! 🎉  I've added a new prop called `persist`, so this way it's opt-in to use the store for caching the products list.

gcsecsey committed:
add option to persist product list in `QueryProductsList` component

cpapazoglou commented on client/components/data/query-products-list/index.jsx:
```diff
@@ -1,9 +1,16 @@
 const request = ( props ) => ( dispatch, getState ) => {
+	if ( props.persist ) {
+		const productsList = getProductsList( getState() );
+		if ( productsList && Object.keys( productsList ).length > 0 ) {
```
I am afraid that the `type` parameter will not allow us to use `Object.keys( productsList ).length > 0` since `type` might be used to fetch a specific segment of products which don't include marketplace (or other) products

https://github.com/Automattic/wp-calypso/blob/6a243bcf5a218680ac6480c919a078e9d2debb7a/client/my-sites/plans/jetpack-plans/selector.tsx#L215

this would result in not having the marketplace products in state and the query not running.

Steps to repro:
- Delete indexed db `productlist` key
- visit `/plans/[DOMAIN]` of a Jetpack site
- visit `/plugins/[DOMAIN]` of WPCOM site

<img width="1590" alt="SS 2022-01-21 at 14 39 49" src="https://user-images.githubusercontent.com/12430020/150528571-513d81ec-c57b-414e-8815-479b9aa70e0b.png">

gcsecsey replied:
I've tried reproducing this with a Free plan as on the screenshot, but the jetpack request isn't fired for me on the `/plans/[DOMAIN]` page. Could you help me with which sites count as a Jetpack site?

Looking at the code where the request should be fired, this is a page where Jetpack plans are listed, right?
https://github.com/Automattic/wp-calypso/blob/980e64f34ca15ebf5afb2242655e0cd2615f8d89/client/my-sites/plans/jetpack-plans/selector.tsx#L215

cpapazoglou replied:
ahh, sorry. I mean self-hosted sites connected to WPCOM via jetpack.

gcsecsey replied:
Thanks for the clarification @cpapazoglou! I was able to reproduce the issue this way. 👍 

I've added a new `type` field to the `productsList` state variable, and when the query is run, the type is being stored there. This way, when determining whether to fetch the query again, we can compare what has been previously saved to state.

gcsecsey committed:
add type field to productsList state var

The type field helps to differentiate which type of list has been persisted previously.

WBerredo reviewed (changes requested):
I tried to reproduce the testing instructions via [direct link](https://calypso.live?image=registry.a8c.com/calypso/app:build-25503), but the product list call is being remade every time I enter or leave the plugin details screen.

gcsecsey committed:
fix selector

gcsecsey commented:
@WBerredo Oops, sorry, the selector was fetching the wrong field name. 😄 

I've tested the latest version on the live environment, it should be fixed now. 🎉

WBerredo reviewed (approved):
LGTM! 🚀

Both paths of testing instructions working well on my side.
